### ISSUE TITLE
Trailing slash removed from URI when finding resource

### ISSUE THREAD

**Issue by neekfenwick** [peej/tonic#191]
I'm not sure if this is a true Issue or not, but my experience with RESTful javascript stores, in the Dojo project dojo/store/JsonRest, it always recommend specifying the target of the store with a trailing slash, e.g. /api/SomeThing/ .. this means that when a query for all objects is done, the request is to /api/SomeThing/ and when a specific resource is requested by ID, the request is to /api/SomeThing/1234.

I find that with tonic, specifying '/api/SomeThing/' as the target results in the Request classes getURIFromEnvironment() function resolving the uri as /api/SomeThing.  This means that when I specify the endpoint in the comment as `@uri /api/SomeThing/` and hit /api/SomeThing/ in a request, it fails to find my Resource, because the Request's uri attribute isn't actually what as in the HTTP request.

Of course this is easy to work around by simply specifying `@uri /api/SomeThing`, and then calling either /api/SomeThing or /api/SomeThing/ both hit that Resource.

Do you think it's buggy, that getURIFromEnvironment() resolves the uri to something other than was in the original request?

**Comment by neekfenwick** [comment]
I have found that since upgrading my installed tonic library to 3.2, I am having to go back over a few systems I built and remove that trailing slash from the `@uri` annotation.  It looks like whatever previous version of tonic I had supported the trailing slash but the newer versions strip it, so it feels like a regression to me.

I'm hoping you don't need me to dig around and find exactly what version I had, it should be easy enough to prove that calling an endpoint with the trailing slash gets it stripped in the way I describe, and I hope you do consider it to be bad behaviour on the part of tonic.  For now, I'll fix my code, but I guess it'll break again if this behaviour is fixed in tonic.

**Comment by peej** [comment]
Since this is what most people want and the trailing slash problem can cause pain, the Tonic Request object takes the pragmatic approach of coercing the request URL to not have a trailing slash. You can see this here https://github.com/peej/tonic/blob/master/src/Tonic/Request.php#L276 when the URL is parsed for language and type information.

If people can think of a use case for differentiating between a URL with and without a trailing slash, then I'm happy to reconsider/accept pull requests.

**Comment by neekfenwick** [comment]
OK, so it's done on purpose, fair enough.  I still feel it's a regression, because I know my code used to work and it broke when upgrading, but you're happy with the current behaviour so I won't argue further.

I notice that the Javascript library I refer to, Dojo, actually has code to handle whether the `target` attribute of the store has a trailing slash or not:

	_getTarget: function(id){
		// summary:
		//		If the target has no trailing '/', then append it.
		// id: Number
		//		The identity of the requested target
		var target = this.target;
		if(typeof id != "undefined"){
			if(target.charAt(target.length-1) == '/'){
				target += id;
			}else{
				target += '/' + id;
			}
		}
		return target;
	},

I think this has changed since I first used the library and ran into the trouble I described.  So, there isn't actually a problem here, I just have to remember to never define my `target` with a trailing slash, which would cause requests without an id to fail.

I'll close this Issue.  Thanks for your time.
